compiler/src/dotty/tools/dotc/typer/Nullables.scala

```diff
@@ -515,7 +548,10 @@ object Nullables:
       && assignmentSpans.getOrElse(sym.span.start, Nil).exists(whileSpan.contains(_))
       && ctx.notNullInfos.impliesNotNull(ref)
 
-    val retractedVars = ctx.notNullInfos.flatMap(_.asserted.filter(isRetracted)).toSet
+    val retractedVars = ctx.notNullInfos.flatMap(info =>
+      if info.asserted == null then Set.empty
+      else info.asserted.filter(isRetracted)
+    ).toSet
     ctx.addNotNullInfo(NotNullInfo(Set(), retractedVars))
```

Comment:
Here we lose knowledge about the loop being unreachable. The result always has `asserted == Set()`, even if it was `null` before.

Reply:
I think this is fine for the NotNullInfo in the `ctx`. Suppose there is a terminated info in the `ctx`, adding a new non-terminated info will not change its behaviour: still treating all symbols as non-nullable.